store/store.go

```diff
@@ -0,0 +1,212 @@
+
+	go db.itemHandler()
+	go db.republishHandler()
+
+	return db, db.ch
+}
+
+func (db *Database) setReplicate() {
+	db.replicate.Lock()
+	db.replicate.time = time.Now().Add(time.Duration(db.tReplicate))
+	db.replicate.Unlock()
+}
+
+func (db *Database) getReplicate() time.Time {
+	db.replicate.RLock()
+	time := db.replicate.time
+	db.replicate.RUnlock()
+
+	return time
+}
+
+// TODO: Max 1000 chars. Truncate input string.
+func (db *Database) AddItem(value string, origPub NodeID) error {
+	t := time.Now()
+	expire := t.Add(db.tExpire)
+	republish := t.Add(db.tRepublish)
+
+	key := Key(blake2b.Sum256([]byte(value)))
+
+	newItem := item{}
+
+	newItem.value = value
+	newItem.expire = expire
+	newItem.republish = republish
+	newItem.origPub = origPub
+
+	// TODO: Handle duplicate keys.
+	db.items.Lock()
+	db.items.m[key] = newItem
+	db.items.Unlock()
+
+	return nil
+}
+
+func (db *Database) AddLocalItem(key Key, value string) {
+	t := time.Now()
+
+	newLocalItem := localItem{}
+
+	newLocalItem.value = value
+	newLocalItem.repubTime = t.Add(db.tRepublish)
+
+	db.localItems.Lock()
+	db.localItems.m[key] = newLocalItem
+	db.localItems.Unlock()
+}
+
+func (db *Database) GetItem(key Key) (reqItem item, err error) {
+	db.items.RLock()
+	requestedItem, found := db.items.m[key]
+	db.items.RUnlock()
+
+	if !found {
+		err = fmt.Errorf("store: GetItem, no item matching key: %x", key)
+		return
+	}
+
+	return requestedItem, nil
+}
+
+/*
+func (db *Database) GetRepubTime(key Key) (repubTime time.Time, err error) {
+	db.localItems.RLock()
+	localItem, found := db.localItems.m[key]
+	db.localItems.RUnlock()
+
+	if !found {
+		err = fmt.Errorf("store: GetRepubTime, no time matching key: %x", key)
+		return
+	}
+
+	return localItem.repubTime, nil
+}
+*/
+
+func (db *Database) GetLocalItem(key Key) (reqLocalItem localItem, err error) {
+	db.localItems.RLock()
+	localItem, found := db.localItems.m[key]
+	db.localItems.RUnlock()
+
+	if !found {
+		err = fmt.Errorf("store: GetLocalItem, no localItem matching key: %x", key)
+		return
+	}
+
+	return localItem, nil
+}
+
+func (db *Database) evictItem(key Key) {
+	db.items.Lock()
+	delete(db.items.m, key)
+	db.items.Unlock()
+}
+
+// Start this as a Goroutine at node start.
+func (db *Database) itemHandler() {
+	for {
+		timer := time.NewTimer(time.Second * 1)
+		<-timer.C
+
+		now := time.Now()
+		var evictees []Key
+
+		db.items.RLock()
+		for key, item := range db.items.m {
+			if now.After(item.expire) || now.After(item.republish) {
+				evictees = append(evictees, key)
+			}
+		}
+		db.items.RUnlock()
+
+		for _, key := range evictees {
+			db.evictItem(key)
+		}
+	}
+}
+
+// Start as Goroutine.
+// Function is responsible for republishing data that should persist on storage network.
+func (db *Database) republishHandler() {
+	for {
+		timer := time.NewTimer(time.Second * 1)
+		<-timer.C
+
+		now := time.Now()
+		replicate := now.After(db.getReplicate()) // This line is a suspect if something doesnt work.
```

Comment:
Why is it suspect?

Reply:
It was a suspect of the data race before. I had changed this many times and lost track of what i thought should be there. :eyes: